### PR TITLE
Add support for tracking pre-release addon versions from Curse

### PR DESCRIPTION
**Pull request by amas0**
This pull request addresses #149 and #151. 

The goal of this enhancement is to support the tracking of pre-release addon versions (alpha/beta) from Curse through an optional setting in `addons.txt`. This is described in the updated README, but the now available syntax allows for addons to be specified like so:

```
https://www.curseforge.com/wow/addons/deadly-boss-mods beta
https://www.curseforge.com/wow/addons/auctionator alpha
```

The AddonManager will split off space-separated text and parse it for alpha/beta strings. The rest of the URL is parsed as before, including subfolder support. This setting is entirely optional and if it is misspecified or omitted, the behavior will fall back on tracking official releases like the tool currently supports.

Powering this new functionality is a refactoring of the logic used to scrape addon versions from Curse. The current master relies on scraping the addon homepage and using regex to find the current releases. This PR instead uses beautifulsoup based parsing of the `/files/all` page of an addon. The new `Curse.versions()` method returns a generator which yields `CurseAddonVersion` objects including all the metadata necessary for comparing versions and getting download links. As a result, the `Curse.find_zip_url()` and `Curse.get_latest_version()` methods are greatly simplified.

The more explicit parsing of version info allows for an easy resolution to #151 as we no longer have to rely on checking whether there are two releases on the main addon page and assuming the first is for retail and the second is for classic.

The changes include a number of new unit tests to validate this functionality, as well as including a mocked up `.html` file in the testing resources directory to ensure correct behavior on an example Curse page.

Looking forward to any comments or requests for change if necessary!

### CODE REVIEW

**Comment by grrttedwards** [comment]
Seems like a great improvement, I'll get time to review this hopefully before the weekend is up! Thanks.

**Review by grrttedwards — CHANGES_REQUESTED** [review]
Overall looks good. Left some minor things I suggested changes for.

- [ ] There is one issue with the following addon, with no release track specified:
https://www.curseforge.com/wow/addons/mythic-dungeon-tools

With configuration:
```
[WOW ADDON UPDATER]
WoW Addon Location = C:\Program Files (x86)\World of Warcraft\_retail_\Interface\AddOns
Addon List File = addons-retail.txt
Installed Versions File = installed-retail.ini
# retail or classic
Game Version = retail
```

and entry:
```
https://www.curseforge.com/wow/addons/method-dungeon-tools
```

The new paginating code fails to find a table:

```
2020-09-07 15:10:23,817 [DEBUG] [12872] [urllib3.connectionpool] https://www.curseforge.com:443 "GET /wow/addons/method-dungeon-tools/files/all?filter-game-version=1738749986%3A517&page=1 HTTP/1.1" 404 None
2020-09-07 15:10:23,861 [ERROR] [12872] [updater.manager.addon_manager] Failed to find addon version number for game version: GameVersion.retail, https://www.curseforge.com/wow/addons/method-dungeon-tools
Traceback (most recent call last):
  File "C:\Users\Garrett\code\wow-addon-updater\updater\site\curse.py", line 102, in versions
    _, *version_rows = versions_table.find_all('tr')
AttributeError: 'NoneType' object has no attribute 'find_all'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\Garrett\code\wow-addon-updater\updater\manager\addon_manager.py", line 100, in update_addon
    latest_version = site.get_latest_version()
  File "C:\Users\Garrett\code\wow-addon-updater\updater\site\curse.py", line 123, in get_latest_version
    latest_release = next(version for version in self.versions() if version.type >= self.addon_version)
  File "C:\Users\Garrett\code\wow-addon-updater\updater\site\curse.py", line 123, in <genexpr>
    latest_release = next(version for version in self.versions() if version.type >= self.addon_version)
  File "C:\Users\Garrett\code\wow-addon-updater\updater\site\curse.py", line 111, in versions
    raise self.version_error() from e
updater.site.abstract_site.SiteError: Failed to find addon version number for game version: GameVersion.retail, https://www.curseforge.com/wow/addons/method-dungeon-tools
```

File: test/site/test_curse.py [hunk 1]
```diff
@@ -1,16 +1,44 @@
...
+                    version_regex=r"[0-9]+\.[0-9]+\.[0-9]+",
+                    supported_game_versions=ALL_VERSIONS),
+    VersionTestData(url='https://www.curseforge.com/wow/addons/weakauras-2',
+                    version_regex=r"[0-9]+\.[0-9]+\.[0-9]+",
+                    supported_game_versions=ALL_VERSIONS)
 ]
 
+MOCK_VERSION_PAGE = MagicMock()
+with open(get_file('mock-curse-version-addons-simc.html'), 'r') as f:
+    MOCK_VERSION_PAGE.text = f.read()
```
[the remark this answers is not in the document]
Comment (amas0) [reply test/site/test_curse.py:40]: I totally agree on the greater point about integration testing with the site itself. The reason I set up _this_ particular static test was to test the `AddonType` ordering. As addons are continuously putting out new versions, it was impossible to consistently test the situation where an addon had alpha, beta, and release version on the same page in that respective order. 

Since the other unit tests should fail if the page format changes, this one more serves as "given the _expected_ html, test that the html -> structured format works properly"

I'm open to modifying how the tests work and there is probably a cleaner way to do so, but there are definitely limitations to testing different edge cases on live data. Happy to explore any suggestions you have!

File: updater/site/curse.py [hunk 2]
```diff
@@ -23,40 +69,60 @@ class Curse(AbstractSite):
...
-            # if classic, choose the explicit "classic version" listed, or fall back to the only version available
-            return classic_version[-1] if self.game_version is GameVersion.classic and classic_version else main_version
+            p = Curse.session.get(f'{self.url}/files/all', params=request_params)
+            soup = bs4.BeautifulSoup(p.text, 'html.parser')
+            versions_table = soup.find('table', {'class': 'listing listing-project-file project-file-listing b-table b-table-a'})
+            # Header row consumed by _
+            _, *version_rows = versions_table.find_all('tr')
+            yield from (CurseAddonVersion.from_tr(row) for row in version_rows)
+
+            # determine if there are more pages of versions, recurse if so
```
[the remark this answers is not in the document]
Comment (amas0) [reply updater/site/curse.py:105]: Given that the logic for pulling the latest version prefers official releases over beta versions over alpha versions -- the only situation where multiple requests to Curse should happen is if an addon has a full page of alpha+beta releases (looks like 25) when a user is tracking official releases.

My bet is that this would be rare enough not to worry about

File: test/site/test_curse.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (grrttedwards) [reply test/site/test_curse.py:40]: That's totally reasonable to me. Love the explanation. Thanks.

File: updater/site/enum.py [hunk 3]
```diff
@@ -1,7 +1,13 @@
...
 class GameVersion(Enum):
     agnostic = auto()
     retail = auto()
     classic = auto()
+
+
+class AddonVersion(IntEnum):
+    release = 3
+    beta = 2
+    alpha = 1
```
[the remark this answers is not in the document]
Comment (amas0) [reply updater/site/enum.py:13]: `IntEnum` supports an explicit ordering of its members whereas `Enum` does not. I used it to make the logic `version.type >= self.addon_version` work when searching for the latest version.

**Comment by amas0** [comment]
Regarding method-dungeon-tools -- this is an odd one because it URL redirects to mythic-dungeon-tools when you navigate to the home page. If you add mythic-dungeon-tools instead, everything works properly.

So the code I wrote breaks if you use the old name of since-renamed addon. I have a few thoughts on options to address this, I want to get your opinion on which would be most appropriate:

1) Do nothing. Once the addon is renamed, there should be an expectation that the user will specify it with the new name in their own addon list. This will break existing installs, which should probably be avoided. But it's also the easiest :)
2) Add in a very specific resolution for _just_ `method-dungeon-tools` -> `mythic-dungeon-tools`. It's not exactly elegant, but I imagine this particular addon is the most prominent situation of an addon being renamed. It's a quick fix, but provides no general solution.
3) Add some sort of logic to attempt to resolve any URL redirects before trying to pull the addon. This would serve as a general solution, but possibly add more calls to Curse than would be ideal. Perhaps it could be done as part of fallback behavior or exception handling to limit the number of calls.

Curious to hear your thoughts. Appreciate the comments!

File: updater/site/enum.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (grrttedwards) [reply updater/site/enum.py:13]: Ah I see. I like it. The docs say it best `IntEnum: Base class for creating enumerated constants that are also subclasses of int.`

**Comment by grrttedwards** [comment]
Oh! Didn't notice the URL redirect. 😉 

I'm not super in love with 1, definitely can't approve 2, 

> 3. Add some sort of logic to attempt to resolve any URL redirects before trying to pull the addon. This would serve as a general solution, but possibly add more calls to Curse than would be ideal. Perhaps it could be done as part of fallback behavior or exception handling to limit the number of calls.

This is actually already implemented for the purpose of "upgrading" old-style Curse URLs, so it could be worth it just to remove the predicate and rename it to something like `_normalize_curse_urls`.

https://github.com/grrttedwards/wow-addon-updater/blob/master/updater/site/curse.py#L60-L72
```
    @classmethod
    def _convert_old_curse_urls(cls, url: str) -> str:
        if any(old_url in url for old_url in [Curse._OLD_URL, Curse._OLD_PROJECT_URL]):
            try:
                # Some old URL's may point to nonexistent pages. Rather than guess at what the new
                # name and URL is, just try to load the old URL and see where Curse redirects us to.
                page = Curse.session.get(url)
                page.raise_for_status()
                return page.url
            except Exception as e:
                raise SiteError(f"Failed to find the current page for old URL: {url}") from e
        else:
            return url
```

File: updater/site/curse.py [hunk 4]
```diff
@@ -23,40 +69,60 @@ class Curse(AbstractSite):
...
-            page = Curse.session.get(self.url)
-            page.raise_for_status()  # Raise an exception for HTTP errors
-            content_string = str(page.content)
-            main_zip_url, *classic_zip_url = re.findall(
-                r"cf-recentfiles-credits-wrapper ml-auto my-auto.+?href=\"(?P<download>.+?)\"",
-                content_string)
-            # if classic, choose the explicit "classic download" listed, or fall back to the only download available
-            zip_url = classic_zip_url[-1] if self.game_version is GameVersion.classic and classic_zip_url else main_zip_url
-            return f'https://www.curseforge.com{zip_url}/file'
+            latest_release = next(version for version in self.versions() if version.type >= self.addon_version)
```
[the remark this answers is not in the document]
Comment (amas0) [reply updater/site/curse.py]: Addressed by 50e32a8e8586b7baba9fa2ea3e7a83e7e5962d3a.

**Comment by amas0** [comment]
Good idea on resolving it in `Curse._convert_old_curse_urls` -- I removed the conditional and it resolved the issue.

My one hesitation on pushing that change is that it would add an extra Curse call to every attempt to update a Curse addon. I think there's some middle ground where full normalization happens in the predetermined cases of "old" URLs or as a first step of exception handling. That way the extra call only happens if strictly necessary. Perhaps it could be worthwhile to also provide a warning to the user indicating that they are using an old URL? 

If you're cool with that extra Curse call, I can go ahead and push the change. Otherwise I'm going to think a bit on an alternative.

**Comment by grrttedwards** [comment]
I like the idea of warning users matching old URLs, good one. You don't have to add that, unless you'd like to.

And as for the extra Curse requests, this is already a very unscientific project, so I don't think one more request is going to hurt 😝

Feel free to think about it as you said, and if you have a better solution, I'm all ears.

Thanks for the time you're spending on this tool, I'm glad it gets some love still!

**Comment by amas0** [comment]
I just went ahead and committed the change with the extra request -- not too much overhead that's worth worrying about I'd say.

I'm going to throw up an issue about warning on old/redirected URLs but I'd rather get this PR merged as is and worry about other enhancements later :) 

Happy to contribute! I stumbled upon this while looking for an addon solution while playing on Linux and it worked great! Wanted to show some love back, thanks for maintaining it

**Comment by grrttedwards** [comment]
Yup. Looks good to me. I'm going to merge to develop and prep a minor version release, probably tomorrow. Thanks

**Review by grrttedwards — APPROVED** [review]
